Do not import private keys to wallets with private keys disabled

achow101 opened the pull request:
Fixes a bug where private keys could be imported to wallets with private keys disabled. Now every RPC which can import private keys checks for whether the wallet has private keys are disabled and errors if it is. Also added an belt-and-suspenders check to `AddKeyPubkeyWithDB` to have it assert that the wallet has private keys enabled.

Sjors reviewed (commented):
Concept ACK. Travis is unhappy. On my own machine `test/functional/wallet_disableprivatekeys.py` failed when it was part of the full test suite, but not when run individually. Odd...

<img width="1433" alt="schermafbeelding 2019-01-23 om 11 50 58" src="https://user-images.githubusercontent.com/10217/51602477-5bfc3c00-1f07-11e9-907f-4a39ce4f3ece.png">

`sethdseed` should also have this check, since it currently throws a confusing "Cannot set a HD seed on a non-HD wallet."

Maybe add the same check to `addmultisigaddress`, since the help says "This functionality is only intended for use with non-watchonly addresses."?

ab677ed looks good otherwise

Suggest tagging this for v0.18 and 0.17 backport.

By the way, we never elaborated on `disableprivatekey` in the [release notes](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.17.0.md).

achow101 commented:
The test failure is running `wallet_disableprivatekeys.py` with `--usecli`. It seems that `bitcoin-cli` is erroring, giving: `error: Error parsing JSON:[{'timestamp': 'now', 'scriptPubKey': {'address': 'n2xmfzEaZi2XyjFYCXxfMbUHyKGk2PaL2k'}, 'keys': ['cQdMcowhL91CXuSRs7FSaujMPNArV8dLZcjq16kHDSDgNaAVTB99']}]`

Sjors commented:
So that's the `importmulti` call? Check if `CRPCConvertParam` in `src/rpc/client.cpp` contains all the non-string params; I think there's a bunch of RPC calls where that's not up to date.

achow101 committed:
tests: unify RPC argument to cli argument conversion and handle dicts and lists

When running tests with --usecli, unify the conversion from argument objects to
strings using a new function arg_to_cli(). This fixes boolean arguments when
using named arguments.

Also use json.dumps() to get the string values for arguments that are dicts and
lists so that bitcoind's JSON parser does not become confused.

achow101 commented:
I fixed the test failure. It's because `--usecli` did not properly handle arguments that are dicts and lists.

achow101 commented:
> `sethdseed` should also have this check, since it currently throws a confusing "Cannot set a HD seed on a non-HD wallet."

Done

> Maybe add the same check to `addmultisigaddress`, since the help says "This functionality is only intended for use with non-watchonly addresses."?

I don't think that's necessary. IMO we should only have this warning for anything where private keys can be imported.

(Missed these in the earlier push)

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -549,6 +552,10 @@ UniValue importwallet(const JSONRPCRequest& request)
         );
 
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Importing wallets is disabled when private keys are disabled");
```
Done

laanwj replied:
Please open an issue for this as well, a TODO in the source code hasn't historically turned out very effective.

Sjors replied:
Although it says "todo", I see it more as an explanation that this is strictly speaking a bug, and that we should pay attention to this if we  move to descriptor based wallets and (presumably) overhaul the dump and import commands.

Sjors commented on test/functional/test_framework/test_node.py:
```diff
@@ -433,8 +433,8 @@ def batch(self, requests):
-        pos_args = [str(arg).lower() if type(arg) is bool else str(arg) for arg in args]
-        named_args = [str(key) + "=" + str(value) for (key, value) in kwargs.items()]
+        pos_args = [str(arg).lower() if type(arg) is bool else (json.dumps(arg) if (isinstance(arg, dict) or isinstance(arg, list)) else str(arg)) for arg in args]
+        named_args = [str(key) + "=" + (json.dumps(value) if (isinstance(value, dict) or isinstance(value, list)) else str(value)) for (key, value) in kwargs.items()]
```
paging @jnewbery, looks ok?

jnewbery replied:
Definitely agree with pulling this monstrosity out into a function :)

bools in named args were fixed here: https://github.com/bitcoin/bitcoin/pull/14938#discussion_r243390546 but that PR wasn't merged.

achow101 replied:
I've pulled this into a new function which is easier to read. It also fixes the bools in named args issue.

achow101 replied:
Since this issue affects multiple PRs, I've split it into a separate PR: #15301

Sjors commented:
tACK 2fa2a44

Sjors reviewed (approved):
tACK 2fa2a44

gmaxwell commented:
Concept ACK.  (maybe we should have a warning that prints in the debug log if you load a non-private-key wallet with private keys in it, like we do for overlong multisig.)

Sjors commented:
re-tACK 7b70a14 on macOS (without `--enable-debug`, because that will crash until the next rebase)

Sjors reviewed (approved):
re-tACK 7b70a14 on macOS (without `--enable-debug`, because that will crash until the next rebase)

achow101 committed:
Refactor importwallet to extract data from the file and then import

Instead of importing keys and scripts as each line in the file is
read, first extract the data then import them.

achow101 committed:
Do not import private keys to wallets with private keys disabled

achow101 commented:
I've gotten rid of the todo in `importwallet` and just implemented it. The latest push refactors `importwallet` so that the import happens in two steps. First the import file is read and the keys and scripts extracted. Then the check for private keys being imported and private keys disabled is done. Then it imports. This will avoid importing wallet dump file that have any private keys and ensure that nothing is imported at all if the file has private keys.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -613,36 +608,60 @@ UniValue importwallet(const JSONRPCRequest& request)
+            }
+            if (time > 0) {
+                pwallet->m_script_metadata[id].nCreateTime = time;
+                nTimeBegin = std::min(nTimeBegin, time);
```
Shouldn't this be `birthtime`? `int64_t birth_time = DecodeDumpTime(vstr[1])` (because all script timestamps are 0 in the dump)

achow101 replied:
No because `birth_time` just becomes `time` in this loop.

Sjors replied:
Ok, but then what's causing the full rescan?

Just tried on `HEAD~2` and getting the same thing, so I'm blaming my own wallet. Just tried this again with a fresh wallet that only has one transaction. Now the rescan is a modest 10623 blocks (I guess that's a built in safety margin?).

 I also got a ton of `Skipping import of ... (script already present)` messages both my original weird wallet and on the fresh one. That doesn't make sense to me, because I'm importing into a brand new wallet.  But it's not new behavior.

achow101 replied:
I'm not sure what might be causing that problem. However the `Skipping import` messages for scripts is expected since segwit scripts are in the dump, but already automatically imported when the keys were imported.

Sjors reviewed (commented):
I'm not sure if d9d6bed is working correctly, or if I'm seeing an already existing bug in `dumpwallet`:

When I dump a wallet that I created a few months ago, the oldest key in the dump file is `2018-08-01T14:51:24Z`, but all the scripts have `0` in the time column. If I then import that into a fresh wallet, it triggers a rescan from genesis.

The line `if (time > 0)` should prevent `nTimeBegin` from becoming 0 this way, but see my comment where I think this goes wrong.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -613,36 +608,60 @@ UniValue importwallet(const JSONRPCRequest& request)
             }
         }
         file.close();
+        for (unsigned int i = 0; i < keys.size(); ++i) {
```
Done

Sjors commented:
tACK 2911ef603

Sjors commented:
re-tACK 2911ef6
